spd/clustering/merge.py

```diff
@@ -0,0 +1,295 @@
+    # ==================================================
+    # compute coactivations
+    activation_mask_orig: Float[Tensor, "samples c_components"] | None = (
+        activations > merge_config.activation_threshold
+        if merge_config.activation_threshold is not None
+        else activations
+    )
+    coact: Float[Tensor, "c_components c_components"] = (
+        activation_mask_orig.float().T @ activation_mask_orig.float()
+    )
+
+    # check shapes
+    c_components: int = coact.shape[0]
+    assert coact.shape[1] == c_components, "Coactivation matrix must be square"
+    assert activation_mask_orig.shape[1] == c_components, (
+        "Activation mask must match coactivation matrix shape"
+    )
+
+    # for speed, we precompute whether to pop components and which components to pop
+    do_pop: bool = merge_config.pop_component_prob > 0.0
+    if do_pop:
+        iter_pop: Bool[Tensor, " iters"] = (
+            torch.rand(merge_config.iters, device=coact.device) < merge_config.pop_component_prob
+        )
+        pop_component_idx: Int[Tensor, " iters"] = torch.randint(
+            0, c_components, (merge_config.iters,), device=coact.device
+        )
+
+    # start with an identity merge
+    current_merge: GroupMerge
+    if initial_merge is not None:
+        current_merge = initial_merge
+    else:
+        current_merge = GroupMerge.identity(n_components=c_components)
+
+    # initialize variables for the merge process
+    k_groups: int = c_components
+    current_coact: Float[Tensor, "k_groups k_groups"] = coact.clone()
+    current_act_mask: Bool[Tensor, "samples k_groups"] = activation_mask_orig.clone()
+    i: int = 0
+
+    # variables we keep track of
+    merge_history: MergeHistory = MergeHistory.from_config(
+        config=merge_config,
+        c_components=c_components,
+        component_labels=component_labels,
+        sweep_params=sweep_params,
+        wandb_url=wandb_run.url if wandb_run else None,
+    )
+
+    # free up memory
+    if not do_pop:
+        del coact
+        del activation_mask_orig
+        del activations
+        activation_mask_orig = None
+
+    # merge iteration
+    # ==================================================
+    # while i < merge_config.iters:
+    pbar = tqdm.tqdm(range(merge_config.iters), unit="iteration", total=merge_config.iters)
+    for i in pbar:
+        # pop components
+        # --------------------------------------------------
+        if do_pop and iter_pop[i]:  # pyright: ignore[reportPossiblyUnboundVariable]
+            # we split up the group which our chosen component belongs to
+            pop_component_idx_i: int = int(pop_component_idx[i].item())  # pyright: ignore[reportPossiblyUnboundVariable]
+            components_in_pop_grp: int = int(
+                current_merge.components_per_group[  # pyright: ignore[reportArgumentType]
+                    current_merge.group_idxs[pop_component_idx_i].item()
+                ]
+            )
```

Comment:
I don't get why "components_in_pop_grp" is an integer when it's "components" not "component". Oh. I guess it should be n_components_in_pop_grp.